### PR TITLE
Rename Show to ToString and Read to FromString

### PR DESCRIPTION
**Pull request by mlutze**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
I found this interesting: https://github.com/typelevel/cats/issues/932

I don't really agree. However, I do think that maybe the signature of `fromString` should return an `Option` instead of a `Result`. Returning an error message as a string seems to specific. This would mean that we could never change such error messages if programmers come to rely on the exact string. What do you think? ... 

If a programmer wants to know exactly *why* - say - a given string cannot be parsed as a Float32 then he should probably not be using a generic type class but rather be using a custom parser.

File: main/src/library/Core/FromString.flix [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+namespace FromString {
+    class FromString[a] {
+        pub def fromString(x: String): Result[a, String]
```
Comment (magnus-madsen) [inline main/src/library/Core/FromString.flix:19]: I think we should change the return type to `Option[a]` per previous discussion.

File: main/src/library/Core/FromString.flix [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+namespace FromString {
+    class FromString[a] {
```
Comment (magnus-madsen) [inline main/src/library/Core/FromString.flix:18]: We need to find a standard way to describe type classes. Here is a first attempt:

```
///
/// A type class for types that can be constructed from strings.
///
```

File: main/src/library/Core/FromString.flix (hunk 1 above)
Comment (magnus-madsen) [reply main/src/library/Core/FromString.flix:19]:
```
///
/// Optionally returns the value associated with the given string `s`.
///

File: main/src/library/Core/ToString.flix [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+namespace ToString {
+    class ToString[a] {
```
Comment (magnus-madsen) [inline main/src/library/Core/ToString.flix:18]: ///
/// A type class for types that can be converted to strings.
///

File: main/src/library/Core/ToString.flix [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        pub def toString(x: a): String
```
Comment (magnus-madsen) [inline main/src/library/Core/ToString.flix:19]: ///
/// Returns a string representation of the given `x`.
///

File: main/src/library/Core/FromString.flix [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    instance FromString[Int64] {
+        def fromString(x: String): Result[Int64, String] = Int64.fromString(x)
+    }
+
+    instance FromString[BigInt] {
+        def fromString(x: String): Result[BigInt, String] = BigInt.fromString(x)
+    }
+
+    instance FromString[Float32] {
```
Comment (magnus-madsen) [inline main/src/library/Core/FromString.flix:46]: Can I ask you to move the floats before the ints. Our informal ordering always goes:

FloatX
IntX
BigInt

File: main/src/library/Core/ToString.flix [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    instance ToString[Float32] {
+        def toString(x: Float32): String = Float32.toString(x)
+    }
+
+    instance ToString[Float64] {
+        def toString(x: Float64): String = Float64.toString(x)
+    }
+
```
Comment (magnus-madsen) [inline main/src/library/Core/ToString.flix:64]: Note to self: Ill have a closer look at these after merge.

**Comment by magnus-madsen** [comment]
Very nice. I am excited to get this merged in. This means that we can very soon got proper printing of lists! :)

**Comment by magnus-madsen** [comment]
A good next step- as the first real test- would be to make string interpolation call into the type class.

**Comment by mlutze** [comment]
> I found this interesting: [typelevel/cats#932](https://github.com/typelevel/cats/issues/932)

There are some fair points there about the use of `read`. But I think it's worth having a generic way to fulfill the very common task of converting simple strings to proper types.

> This would mean that we could never change such error messages if programmers come to rely on the exact string. What do you think?

Yes, I agree. I think, as a rule, we should never give error messages to the user in such a way that they could/would match on them. If there were more than one possible error here, I'd propose some error enum, but since NumberFormatException is the only thing that could go wrong (I think), it makes sense just to return an Option.

File: main/src/library/Core/FromString.flix (hunk 1 above)
Comment (mlutze) [reply main/src/library/Core/FromString.flix:19]: Should we change the respective library functions as well?

**Comment by magnus-madsen** [comment]
Ready for merge?

**Comment by mlutze** [comment]
> Ready for merge?

Ready!

**Pull request merged**